### PR TITLE
gen4: outer joins

### PR DESCRIPTION
**Pull request by systay**
## Description
Teaches gen4 how to plan outer joins.

The big change is that we had to introduce a new abstraction - the operator tree.
Before this PR, gen4 turned queries into query graphs for planning. That is not enough to handle outer joins well, so we introduced the operator tree, which can be or contain a querygraph to represent the query.

## Related Issue(s)
#7280

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 1]
```diff
@@ -275,446 +294,18 @@ func planHorizon(sel *sqlparser.Select, plan logicalPlan, semTable *semantics.Se
...
-			lhsOffset++
-		} else {
-			jp.columns = append(jp.columns, rhsOffset)
-			rhsOffset--
-		}
-	}
-	return resultIdx
 }
 
 func pushPredicate2(exprs []sqlparser.Expr, tree joinTree, semTable *semantics.SemTable) (joinTree, error) {
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/route_planning.go:311]: nit: rename to pushJoinPredicate

File: go/vt/vtgate/planbuilder/abstract/querygraph.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+		Alias      *sqlparser.AliasedTableExpr
+		Table      sqlparser.TableName
+		Predicates []sqlparser.Expr
+	}
+)
+
+// PushPredicate implements the Operator interface
+func (qg *QueryGraph) PushPredicate(expr sqlparser.Expr, semTable *semantics.SemTable) error {
+	for _, e := range sqlparser.SplitAndExpression(nil, expr) {
+		err := qg.collectPredicate(e, semTable)
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/abstract/querygraph.go:60]: I tried, and the test for this query started failing:
```"select user.col from user join user_extra on user.id = 5 and user.id = user_extra.user_id"```

File: go/vt/sqlparser/analyzer.go [hunk 3]
```diff
@@ -323,6 +323,35 @@ func SplitAndExpression(filters []Expr, node Expr) []Expr {
...
+		return exprs[0]
+	default:
+		result := (Expr)(nil)
+		for i, expr := range exprs {
+			if result == nil {
+				result = expr
+			} else {
+				found := false
+				for j, ex := range exprs {
+					if i != j && EqualsExpr(expr, ex) {
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/sqlparser/analyzer.go]: there was also a bug in this check. I have added a test case for it and fixed it.

**Review by harshit-gangal — APPROVED** [review]

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/vtgate/planbuilder/abstract/join.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		j.Exp = sqlparser.AndExpressions(j.Exp, expr)
+		return nil
+	}
+
+	return semantics.Gen4NotSupportedF("still not sure what to do with this predicate")
+}
+
+// TableID implements the Operator interface
+func (j *Join) TableID() semantics.TableSet {
+	return j.RHS.TableID().Merge(j.LHS.TableID())
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/abstract/join.go:48]: We will look at this optimization later, if needed.

**Pull request merged**